vpr/src/analytical_place/full_legalizer.cpp

```diff
@@ -281,6 +291,710 @@ static LegalizationClusterId create_new_cluster(PackMoleculeId seed_molecule_id,
+                                              block_type_utils,
+                                              vpr_setup_.PackerOpts.target_device_utilization,
+                                              vpr_setup_.PackerOpts.device_layout);
+        // Exit if clusters fit on device.
+        if (fits_on_device)
+            break;
+
+        // Destroy the orphan window clusters to recreate with bigger search radius.
+        VTR_LOG("Clusters did not fit on device with orphan window search radius of %d.\n", orphan_window_search_radius);
+        for (LegalizationClusterId cluster_id: orphan_window_clusters) {
+            if (!cluster_id.is_valid()) continue;
+            cluster_legalizer.destroy_cluster(cluster_id);
+        }
+        orphan_window_clusters.clear();
+    }
+
+    if (!fits_on_device) {
+        VPR_FATAL_ERROR(VPR_ERROR_AP, "Created clusters could not fit on device.");
+    }
+
+    // Report the clustering summary.
+    report_clustering_summary(cluster_legalizer,
+                              neighbor_pass_molecules,
+                              orphan_window_clusters);
+
+    // Check and output the clustering.
+    cluster_legalizer.compress();
+    std::unordered_set<AtomNetId> is_clock = alloc_and_load_is_clock();
+    check_and_output_clustering(cluster_legalizer, vpr_setup_.PackerOpts, is_clock, &arch_);
+
+    // Reset the cluster legalizer. This is required to load the packing.
+    cluster_legalizer.reset();
+    // Regenerate the clustered netlist from the file generated previously.
+    // FIXME: This writing and loading from a file is wasteful. Should generate
+    //        the clusters directly from the cluster legalizer.
+    vpr_load_packing(vpr_setup_, arch_);
+
+    // Verify the packing
+    check_netlist(vpr_setup_.PackerOpts.pack_verbosity);
+    writeClusteredNetlistStats(vpr_setup_.FileNameOpts.write_block_usage);
+}
+
+void FlatRecon::place_clusters(const PartialPlacement& p_placement) {
+    // Setup the global variables for placement.
+    g_vpr_ctx.mutable_placement().init_placement_context(vpr_setup_.PlacerOpts, arch_.directs);
+    g_vpr_ctx.mutable_floorplanning().update_floorplanning_context_pre_place(*g_vpr_ctx.placement().place_macros);
+
+    // The placement will be stored in the global block loc registry.
+    BlkLocRegistry& blk_loc_registry = g_vpr_ctx.mutable_placement().mutable_blk_loc_registry();
+
+    // Create the noc cost handler used in the initial placer.
+    std::optional<NocCostHandler> noc_cost_handler;
+    if (vpr_setup_.NocOpts.noc)
+        noc_cost_handler.emplace(blk_loc_registry.block_locs());
+
+    // Create the RNG container for the initial placer.
+    vtr::RngContainer rng(vpr_setup_.PlacerOpts.seed);
+
+    // Cast the partial placement to flat placement here. So that it can be
+    // used to guide the initial placer and for logging results. This enables
+    // the flow to be used with direct output of GP as well.
```

Comment:
Explain p_placement has been set by the GP (if using internal VTR AP), or by reading in the flat placement file. Cast / copy it to the flat_placement data structures so we can always use them.